app/main/views/suppliers.py

```diff
@@ -191,10 +149,200 @@ def update_supplier_name(supplier_id):
+
+    if form.validate_on_submit():
+        if form.companies_house_number.data:
+            update_supplier_payload = {
+                "companiesHouseNumber": form.companies_house_number.data.upper(),
+                "otherCompanyRegistrationNumber": None
+            }
+            update_declaration_payload = {
+                "supplierCompanyRegistrationNumber": form.companies_house_number.data.upper()
+            }
+        else:
+            update_supplier_payload = {
+                "companiesHouseNumber": None,
+                "otherCompanyRegistrationNumber": form.other_company_registration_number.data
+            }
+            update_declaration_payload = {
+                "supplierCompanyRegistrationNumber": form.other_company_registration_number.data
+            }
+
+        # Although we've fetched the reg number from the supplier, any recent declaration should be updated too
+        supplier_framework_interests = data_api_client.get_supplier_frameworks(supplier_id)["frameworkInterest"]
+        supplier_frameworks = interesting_frameworks(supplier_framework_interests, current_user, frameworks)
+        _, most_recent_declaration = get_company_details_and_most_recent_interest(supplier_frameworks, supplier)
+
+        # Update supplier
+        data_api_client.update_supplier(
+            supplier_id=supplier_id,
+            supplier=update_supplier_payload,
+            user=current_user.email_address
+        )
+
+        if most_recent_declaration.get('declaration'):
+            data_api_client.update_supplier_declaration(
+                supplier_id,
+                most_recent_declaration['frameworkSlug'],
+                update_declaration_payload,
+                current_user.email_address
+            )
+
+        flash(SUPPLIER_DETAILS_UPDATED_MESSAGE.format(supplier_name=supplier['name']))
+        return redirect(url_for('.supplier_details', supplier_id=supplier_id))
+
+    errors = get_errors_from_wtform(form)
+
+    return render_template(
+        "suppliers/edit_registered_company_number.html",
+        supplier=supplier,
+        form=form,
+        errors=errors,
+    ), 200 if not errors else 400
+
+
+@main.route('/suppliers/<int:supplier_id>/edit/registered-address', methods=['GET', 'POST'])
+@role_required('admin-ccs-data-controller')
+def edit_supplier_registered_address(supplier_id):
+    frameworks = data_api_client.find_frameworks()['frameworks']
+    supplier = data_api_client.get_supplier(supplier_id)['suppliers']
+    contact = supplier["contactInformation"][0]
+
+    # Get SupplierFrameworks for frameworks the role is interested in, sorted by oldest frameworkLiveAtUTC first
+    supplier_framework_interests = data_api_client.get_supplier_frameworks(supplier_id)["frameworkInterest"]
+    supplier_frameworks = interesting_frameworks(supplier_framework_interests, current_user, frameworks)
+
+    # Get the company details (either from supplier or recent declaration)
```

Comment:
```suggestion
    # Get the company details (either from supplier or recent declaration)
    # TODO: handle case where they don't have a declaration better (maybe different form?)
```

Reply:
Actually I think I have a solution for this. Commit incoming...